### PR TITLE
Custom Input Toolbar

### PR DESCRIPTION
**Pull request by antipex**
* Updated to latest JSQMessagesViewController `master` code
* Added support for our custom input toolbar and camera roll image picker

### CODE REVIEW

File: JSQMessagesDemo/DemoMessagesViewController.m [hunk 1]
```diff
@@ -18,6 +18,12 @@
 
 #import "DemoMessagesViewController.h"
```
Comment (esilverberg) [inline JSQMessagesDemo/DemoMessagesViewController.m]: Lets revert the demo edits

File: JSQMessagesViewController/Controllers/JSQMessagesKeyboardController.m [hunk 2]
```diff
@@ -130,7 +130,6 @@ - (void)endListeningForKeyboard
     [self jsq_unregisterForNotifications];
 
     [self jsq_setKeyboardViewHidden:NO];
-    self.keyboardView = nil;
```
Comment (esilverberg) [inline JSQMessagesViewController/Controllers/JSQMessagesKeyboardController.m]: If you undo this deletion does the code still work?

File: JSQMessagesViewController/Controllers/JSQMessagesViewController.m [hunk 3]
```diff
@@ -202,6 +209,17 @@ - (void)jsq_configureMessagesViewController
 
     self.topContentAdditionalInset = 0.0f;
 
+    self.bottomContentAdditionalInset = 0.0f;
+
+    self.isAdVisible = YES;
+
+//    [self setInputToolbarShadowVisible:YES];
```
Comment (esilverberg) [inline JSQMessagesViewController/Controllers/JSQMessagesViewController.m]: remove this comment

File: JSQMessagesViewController/Controllers/JSQMessagesViewController.m [hunk 4]
```diff
@@ -912,6 +934,16 @@ - (void)keyboardController:(JSQMessagesKeyboardController *)keyboardController k
...
 
+    // Hide the picker view if the keyboard/input view is dismissed via the
+    // pan gesture recognizer
+
+    if (heightFromBottom == 0.0 && self.isPickerViewVisible) {
+        self.isPickerViewVisible = NO;
+    }
+
+    // Show/hide the ad container view
+    self.isAdVisible = heightFromBottom == 0.0;
```
Comment (esilverberg) [inline JSQMessagesViewController/Controllers/JSQMessagesViewController.m:943]: Is this sufficient? One must also determine free/pro state for the member, in addition to tricky calculations which determine if a picker is showing and therefore hiding our ad view

File: JSQMessagesViewController/Controllers/JSQMessagesViewController.m [hunk 5]
```diff
@@ -1152,4 +1194,63 @@ - (void)jsq_addActionToInteractivePopGestureRecognizer:(BOOL)addAction
...
 }
 
+#pragma mark - Scruff Additions
+
+- (void)setIsPickerViewVisible:(BOOL)isPickerViewVisible {
+    if (_isPickerViewVisible != isPickerViewVisible) {
+        _isPickerViewVisible = isPickerViewVisible;
+
+        if (isPickerViewVisible) {
+            [self.inputToolbar setHidden:YES];
```
Comment (esilverberg) [inline JSQMessagesViewController/Controllers/JSQMessagesViewController.m]: Just line-by-line comment the thinking

File: JSQMessagesViewController/Controllers/JSQMessagesViewController.m [hunk 6]
```diff
@@ -1152,4 +1194,63 @@ - (void)jsq_addActionToInteractivePopGestureRecognizer:(BOOL)addAction
...
+
+            self.inputToolbar.contentView.textView.customInputView = self.pickerView;
+
+            [self.inputToolbar.contentView.textView becomeFirstResponder];
+
+            UITextInputAssistantItem *item = self.inputToolbar.contentView.textView.inputAssistantItem;
```
Comment (esilverberg) [inline JSQMessagesViewController/Controllers/JSQMessagesViewController.m]: Comment what this is for

File: JSQMessagesViewController/Controllers/JSQMessagesViewController.m [hunk 7]
```diff
@@ -202,6 +209,17 @@ - (void)jsq_configureMessagesViewController
 
     self.topContentAdditionalInset = 0.0f;
 
+    self.bottomContentAdditionalInset = 0.0f;
+
+    self.isAdVisible = YES;
+
+//    [self setInputToolbarShadowVisible:YES];
+
+    self.adContainerView.clipsToBounds = YES;
```
Comment (esilverberg) [inline JSQMessagesViewController/Controllers/JSQMessagesViewController.m]: is this duplicative of the code below? Let's just have in one place if possible.

File: JSQMessagesViewController/Controllers/JSQMessagesViewController.m [hunk 8]
```diff
@@ -1152,4 +1194,63 @@ - (void)jsq_addActionToInteractivePopGestureRecognizer:(BOOL)addAction
...
+
+    self.bottomContentAdditionalInset = targetAdHeight;
+}
+
+// Override in subclass to change the ad height
+- (CGFloat)targetAdHeight {
+    return 44.0;
+}
+
+- (void)setInputToolbarShadowVisible:(BOOL)visible {
```
Comment (esilverberg) [inline JSQMessagesViewController/Controllers/JSQMessagesViewController.m]: let us remove this

File: JSQMessagesViewController/Views/JSQMessagesComposerTextView.m [hunk 9]
```diff
@@ -154,6 +154,8 @@ - (BOOL)canPerformAction:(SEL)action withSender:(id)sender {
         return [UIPasteboard generalPasteboard].string != nil || [UIPasteboard generalPasteboard].image != nil;
     }
 
+    [UIMenuController sharedMenuController].menuItems = nil;
```
Comment (esilverberg) [inline JSQMessagesViewController/Views/JSQMessagesComposerTextView.m]: what is this for?

File: JSQMessagesViewController/Views/JSQMessagesInputToolbar.m [hunk 10]
```diff
@@ -110,7 +110,8 @@ - (void)toggleSendButtonEnabled
     BOOL hasText = [self.contentView.textView hasText];
 
     if (self.sendButtonOnRight) {
-        self.contentView.rightBarButtonItem.enabled = hasText;
+//        self.contentView.rightBarButtonItem.enabled = hasText;
```
Comment (esilverberg) [inline JSQMessagesViewController/Views/JSQMessagesInputToolbar.m]: explain why we switched the logic in a comment

**Review by esilverberg — CHANGES_REQUESTED** [review]

File: JSQMessagesViewController/Controllers/JSQMessagesKeyboardController.m (hunk 2 above)
Comment (antipex) [reply JSQMessagesViewController/Controllers/JSQMessagesKeyboardController.m]: Sure does!

File: JSQMessagesViewController/Controllers/JSQMessagesViewController.m (hunk 4 above)
Comment (antipex) [reply JSQMessagesViewController/Controllers/JSQMessagesViewController.m:943]: This determines if we should show or hide the ad container based on whether or not the keyboard or picker view is showing.

**Pull request closed without merging**